[JuliaLowering] Refactor scope resolution pass

mlechu opened the pull request:
## Fixes

### Soft scope
We currently treat flisp's `'(block (softscope true))` as a scope that "is
soft", but this isn't correct. It should behave more like a toggle, where if the
scope surrounding `(softscope true)` is the top-level thunk, neutral scopes not
protected by a hard scope become permeable to already-defined globals.  I've
added `K"softscope"` for this.

Fixes https://github.com/JuliaLang/JuliaLowering.jl/issues/101.
`JuliaLowering.activate!()` should be much more usable in the REPL now, as
globals won't be accidentally eaten by the "soft scope."

### Shadowing behaviour
Found while cleaning up the lhs-resolution step with the change above.  This PR
allows static parameters to be shadowed by globals and locals as long as they're
explicit (and not in the same scope).  flisp allows this with globals, and the
explicit locals that desugar to `local-def` forms (which JuliaLowering doesn't
have).

This change is more permissive than flisp in the local case, since after looking
into why shadowing was disallowed I realized it was just was just to prevent
assignment to the static parameter (#32623).  The flisp fix leads to some funny
behaviour:

```
julia> function f(x::T) where T
     let
         global T # remove this T and the other two will fight
         let; local T; end
     end
     end
f (generic function with 1 method)
```

### Bindings/LambdaBindings
In figuring out how to use the variable bookkeeping system, I ran into inaccuracies.

`LambdaBindings` is currently a per-lambda map from unique variable to four
flags: `captured`, `read`, `assigned`, and `called`.  I think (but correct me if
I'm wrong @c42f) this was a misinterpretation of the holy text: in flisp
lowering, a local variable captured by some other lambda does show up in both
lambdas' variable lists, but is the same underlying object, and flag mutations
on one variable are seen by all lambdas.

I tried to think of other reasons for tracking vinfo per lambda within lowering,
but if we're doing something about the capture-boxing issue, we need something
more complex anyway.

This PR moves all vinfo to `BindingInfo` and deletes the incorrect bookkeeping
in `LambdaBindings`. We still need to have a per-lambda flag for capturedness
(different from the variable-level `capt` vinfo flag). With the added flags,
I've just made BindingInfo mutable since our previous workflow (BindingId is an
index into a vector; mutate the vector) doesn't give us the benefits of
immutability anyway.

## Enhancements
- Scopes are retained until the end of the pass, so consumers like JETLS can
  answer questions like "what names are available at my cursor?"  Recreating
  this previously-discarded information was a bit hacky! [[1]](https://github.com/aviatesk/JETLS.jl/blob/6b43afe9e6f5392fc63714a8127a7653e94d9dba/src/utils/ast.jl#L115-L200) [[2]](https://github.com/iuliadmtru/ScopeVars/blob/f7da77eb07d644d3bcd57a3677791b6158d2d8b1/src/find_scope_vars.jl)
- Hopefully enough tests and explanatory comments to make up for the large diff

## TODO
- Our use of `K"local"` in desugaring is dubious in some places
- I've added `expr_compat_mode` to the scope analysis context, but we still need
  to implement flisp hygiene exemptions for globals (see note in test/scopes.jl).
- Reviewing the IR changes, the `#self#` argument still has extra flags set in
  some cases.  This is an existing desugaring bug with a comment that took me
  too long to find: we shouldnn't be using the same `#self#` binding for
  multiple methods defined by one function body

mlechu committed:
bikeshed: rename `lookup_binding` to `get_binding`

"get" is clearer to me that the binding is the output rather than the input,
     that it definitely exists, and that the ID resolves uniquely

mlechu committed:
Small fix: `kw` in `function (tuple=1); end` syntax

mlechu committed:
Refactor scope resolution pass

topolarity commented on JuliaLowering/src/bindings.jl:
```diff
@@ -1,74 +1,69 @@
+    binfo.is_assigned_once   && print(io, ", is_assigned_once")
+    binfo.is_captured        && print(io, ", is_captured")
+    binfo.is_always_defined  && print(io, ", is_always_defined")
+    binfo.is_used_undef      && print(io, ", is_used_undef")
```
Doesn't this no longer round-trip when pasting into, e.g., the REPL?

mlechu committed:
`K"use_softscope_if_toplevel"` -> `K"softscope"`

I know it's a goal to make some of these clearer, but using the same string as
     Expr (and using the fall-through case in conversion) will make some planned
     code movement easier

topolarity commented on JuliaLowering/test/scopes.jl:
```diff
@@ -58,23 +58,281 @@ begin
+    # errors like the conflict cases (two of the same decl should never do
+    # anything, and the user might be expecting two variables).
+    @testset "global,global" begin
+        s = "function (); global g; global g; 1; end"
```
JuliaLowering seems to accept this example, but flisp does not:
```julia
julia> JuliaLowering.include_string(TestMod,"""
       function (); global g; const g = 1; end
       """)
#anon###0 (generic function with 1 method)

julia> function (); global g; const g = 1; end
ERROR: syntax: `global const` declaration not allowed inside function around REPL[5]:1
Stacktrace:
 [1] top-level scope
   @ REPL[5]:1
```

somehow `global g` gets JL to look the other way

topolarity commented on JuliaLowering/src/scope_analysis.jl:
```diff
@@ -17,493 +17,405 @@ function NameKey(ex::SyntaxTree)
 end
 
-function _resolve_scopes(ctx, ex::SyntaxTree)
+function _resolve_scopes(ctx, ex::SyntaxTree, @nospecialize(scope))
```
```suggestion
function _resolve_scopes(ctx, ex::SyntaxTree, @nospecialize(scope::Union{Nothing,ScopeInfo}))
```

IIUC you assume this, so best to mark it

topolarity reviewed (commented):
I still need to do another once-over of the core pass logic, but this should be useful (mostly very minor) feedback in the mean time.

topolarity commented on JuliaLowering/src/scope_analysis.jl:
```diff
@@ -17,493 +17,405 @@ function NameKey(ex::SyntaxTree)
+    # Every lexical scope, indexed by ScopeId
+    scopes::Vector{ScopeInfo}
+    # Current stack of scopes to look for names in, innermost scope last
+    scope_stack::Vector{ScopeId}
```
Is this identical to the sequence of scopes you get if you follow the `parent(ctx, scope)` chain?

mlechu replied:
(discussed in meeting) I'm mirroring our handling of macro scope layers but agree doing `parent` would be cleaner.

topolarity commented on JuliaLowering/src/scope_analysis.jl:
```diff
@@ -17,493 +17,405 @@ function NameKey(ex::SyntaxTree)
+    elseif old_k === new_k
+        (new_k === :global || new_k === :local) && return bid
+        throw(LoweringError(ex, "function $(_var_str(new_k)) name not unique"))
+    # See note in test/scopes.jl
```
can you add some identifier to the note you mean?

topolarity commented on JuliaLowering/src/scope_analysis.jl:
```diff
@@ -17,493 +17,405 @@ function NameKey(ex::SyntaxTree)
+        # mutable nameless bindings may be introduced in desugaring
+        maybe_declare_in_scope!(ctx, scope, SyntaxTree(ctx.graph, node_id),
+                                get_binding(ctx, bid).kind)
     end
```
It feels like it should be possible to unify this with the logic for `scope.assignments` below, since semantically they aren't any different iiuc?

Not a required re-work by any means, but it seems like it might be nice

topolarity reviewed (approved):
Seems like the right approach (and indeed closer to flisp) to update shared bindings across all lambdas simultaneously.

Changes are looking good to me. Nice work @mlechu!

Feel free to take or drop the nits / ideas, as desired.

mlechu committed:
Code review comments

Co-authored-by: Cody Tapscott <topolarity@tapscott.me>

mlechu commented on JuliaLowering/src/bindings.jl:
```diff
@@ -106,147 +101,81 @@ function _binding_id(ex::SyntaxTree)
     # Reconstruct the SyntaxTree for this binding. We keep only the node_id
     # here, because that's got a concrete type. Whereas if we stored SyntaxTree
     # that would contain the type of the graph used in the pass where the
     # bindings were created and we'd need to call reparent(), etc.
```
They are pretty similar, but for different purposes. `reparent` is "change the graph of this tree (graph structure must be identical to the original, but attributes on the graph may change)".  `binding_ex(ctx, id)` is just an alternative to an `@ast` call where we don't want to make another node (or don't have a good provenance reference for one).  I'm not sure it works so well as a space-saving measure given my comment above on binding nodes not being reused, though.

mlechu commented on JuliaLowering/src/scope_analysis.jl:
```diff
@@ -17,493 +17,405 @@ function NameKey(ex::SyntaxTree)
+    # assigned to at top level, or passing the defined-and-owned-global check.
+    defined_toplevel_globals::Set{NameKey}
+    enable_soft_scopes::Bool
+    expr_compat_mode::Bool
```
Yup, I may have lost a sentence somewhere in the PR comment.  We'll need this if we want to replicate flisp's bad hygiene with globals with some hope of making things better.

mlechu commented on JuliaLowering/test/scopes.jl:
```diff
@@ -58,23 +58,281 @@ begin
+
+# For each distinct outer and inner scope, and each kind of variable in the
+# outer scope, set the same name to true from the inner scope
+@testset "Behaviour of `=` in local scope (shadow or assign-existing)" begin
```
Gives me a sense of security.  If you have other sledgehammer-testing ideas, send them my way!

c42f commented on JuliaLowering/test/scopes.jl:
```diff
@@ -58,23 +58,292 @@ begin
+
+    # globals may overlap args or sparams (buggy?)  TODO: decide whether it's
+    # worth replicating this behaviour.  We would likely need to copy the way
+    # flisp nests an extra scope block in every lambda.
```
Eek.  I can't recall right now whether I intentionally tried to depart from the weird flisp behavior or not.  seems potentially breaking to be different (if logical) - may need to be changed via syntax evolution I guess.

c42f replied:
Sounds reasonable.

c42f commented on JuliaLowering/test/scopes.jl:
```diff
@@ -58,23 +58,292 @@ begin
+    @testset "sparam,local" begin
+        s = "function (a::s) where {s}; local s; end"
+        @test_throws LoweringError JuliaLowering.include_string(test_mod, s)
+    end
```
Maybe it would also help if we simply added a bit more structure to the IR tests.   Eg, allow structuring the IR test files so that we can group the content into test sets?

c42f replied:
> My main gripe with our tests isn't duplication, it's that we need to write more of them :)

Haha! It's the bane of any new project when switching from "this is going fast and everything is changing all the time" to "oh this about to be production code".

Having done several big new development projects, there's always this transition period, and it's critical to the longer term success. So I really appreciate your efforts - and likewise I'll try to do my best to beef up tests whenever I see any missing.

If I find a way to make the tests "more fun" to write I'll also put in the effort there.  I am happy that we have IR tests at all - in contrast to existing lowering. They're still a bit clunky to work with - perhaps we may yet find a way to improve that.

c42f commented:
As another more concrete example of why `LambdaBindings` is useful, consider the `is_called` flag.

This is metadata which applies to how argument slots are used within the body of the `CodeInfo` - it eventually gets mapped into the CodeInfo's `called` field and used within gf.c for specialization heuristics. These heuristics shouldn't care about the use of the binding in places where it's captured, but merely about its direct use in the body of the code info. I would argue that flisp is wrong if it tracks this information globally. To make this concrete:

```julia
function f(g)
    function h()
        g()  # should not set `is_called` on `g`'s slot flag in `f`
    end
end
```

You're certainly right to notice a disconnect with flisp - because I didn't actually port the flisp implementation of `vinfo` closely, but instead referred to the runtime C code's actual usage of the metadata which is being produced by lowering as the source of truth for functionality.

c42f commented:
> Calling h() inside of f() would suffer (it will not infer well) if g is not marked as called

Hmm. I was convinced it would be the job of the optimizer to deal with inlining the implementation of `h`, after which the call of `g()` would be self-evident... the same behavior as if `h` were a manually defined closure:

```julia
struct _h_type{G}
    g::G
end

function (h::_h_type)()
   h.g()
end

function f(g)
    h = _h_type(g)
    h()   # lowering cannot know that g is called
end
```

But perhaps I'm mistaken and this is a case where the compiler can use the fact that `h` is a closure to do better than the hand written version?

We also have the following example where the heuristic guesses wrong because the closure isn't called. But I'll concede it's a less likely case and it's high probability that any closure will be called during the lifetime of the enclosing function.

```julia
function f(g)
    function h()
        g()
    end
    return h
end
```

c42f commented:
Basically - I admit I thought this was an outright bug which I was fixing but it's not actually simple :sweat_smile:

c42f commented:
> I'll elaborate on my claim "we need something more complex anyway". Almost every lambda-vinfo-flags-based unboxing proof I can think of requires other machinery complex enough to make the flags unnecessary.

I agree that information global to the body of a lambda isn't very powerful. It can help, but the cases it helps with are rather trivial - and I don't think very useful. We need some simple data flow analysis to do unboxig as best we can within lowering.  And "as best we can" is not very powerful because our escape analysis will be extremely pessimistic without information from the optimizer.

Ok, so I'm a bit happier with deleting this stuff for now, assuming that we may need to completely refactor it in the next step anyway.  Would still like to know whether the `is_called` heuristic's behavior with respect to captures was intended or a mere side effect of the way vinfo happens to work currently.

c42f commented:
> Nonetheless, I think we'll want to save that for "Phase 2" of the JuliaLowering initiative, after we have already gotten parity with flisp

I think that's pretty fair in this case - even if it's a bug (I'm now unsure about this) - keeping consistency with flisp is reasonable because changing specialization could have subtle side effects and need its own review.

Side note - there's several other probable-bugs which I came up against in working through the flisp code for the lowering port as a whole. I mostly chose to fix them when they seemed obviously buggy so we'll be able to close a few existing issues once the integration work is done. Obviously we could reimplement the bugs but that's a bit depressing so hopefully we'll be able to get away with just leaving them fixed :sweat_smile:
